usr/sbin/p11sak/p11sak.c

```diff
@@ -2292,11 +2287,47 @@ static CK_RV generate_asymmetric_key(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
         goto done;
     }
 
-    rc = set_battr(attr_string, pub_attr, &pub_acount, prv_attr, &prv_acount);
-    if (rc != CKR_OK) {
-        fprintf(stderr, "Error setting binary attributes (error code 0x%lX: %s)\n", rc,
-                p11_get_ckr(rc));
-        goto done;
+    /**
+     * if the separator sign is not in the string the custom attributes apply for 
+     * public and private key
+     * else the attributes in front of the separator sign are set for the public 
+     * key and the ones behind are set for the private key
+     */
+    char *ret = NULL;
+    if (attr_string) {
+        ret = strchr(attr_string, separator);
+    }
+
+    if (attr_string && ret != NULL) {
+        // setting the attributes after the separator for the private key
+        rc = set_battr(ret + 1, prv_attr, &prv_acount, 1);
+        if (rc != CKR_OK) {
+            fprintf(stderr, "Error setting binary attributes for private key (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            goto done;
+        }
+        // setting the attributes in front of the separator for the public key
+        *ret = '\0';
+        rc = set_battr(attr_string, pub_attr, &pub_acount, 0);
+        if (rc != CKR_OK) {
+            fprintf(stderr, "Error setting binary attributes for public key (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            goto done;
```

Comment:
maybe also do `*ret = separator;` here to restore the separator char also in the error case.

Reply:
Done